### PR TITLE
Feature Application Launcher 3.x

### PR DESCRIPTION
**Pull request by cdcabrera**
Application launcher directive

Currently making use of CSS/Styling from base PatternFly 3.x, PR pending: 

- https://github.com/patternfly/patternfly/pull/632

Samples

- https://rawgit.com/cdcabrera/patternfly/Feature-AppLauncher-3x-dist/dist/tests/application-launcher-nav.html

screen shot 2017-04-19 at 3 54 05 pm

@dtaylor113 @jeff-phillips-18 @serenamarie125

### CODE REVIEW

File: src/navigation/launcher-navigation-directive.js [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+    return {
+
+      restrict: 'A',
+
+      replace: true,
+
+      scope: {
+
+        items: '&',
```
Comment (dtaylor113) [inline src/navigation/launcher-navigation-directive.js]: Think you can get rid of the extra newlines

File: test/navigation/launcher-navigation.spec.js [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+  it('should have hidden application icons', function () {
+
+    var htmlTmp = '<div pf-launcher-navigation items="sites" is-open="false" is-disabled="false" is-list="true" hidden-icons="true"></div>';
+
+    compileHTML(htmlTmp, $scope);
+
+    var content = element.find('[role="menuitem"] .fa');
+
+    expect(content.length).toBe(0);
+  });
```
Comment (dtaylor113) [inline test/navigation/launcher-navigation.spec.js]: Not sure we need a newline between each line.

File: src/navigation/launcher-navigation-directive.js (hunk 1 above)
Comment (cdcabrera) [inline src/navigation/launcher-navigation-directive.js]: can do

File: test/navigation/launcher-navigation.spec.js (hunk 2 above)
Comment (cdcabrera) [inline test/navigation/launcher-navigation.spec.js]: can do

File: src/navigation/launcher-navigation-directive.js [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        isOpen: '=?',
+        isDisabled: '&?',
+        isList: '&?',
+        hiddenIcons: '&?'
```
Comment (jeff-phillips-18) [inline src/navigation/launcher-navigation-directive.js]: Why are these '&' ? Shouldn't they be '=' ?

File: src/navigation/launcher-navigation-directive.js [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        /**
+         * @ngdoc method
+         * @name toggled
+         * @methodOf patternfly.navigation.directive:pfLauncherNavigation
+         * @description
+         * Override and relate open and closed launcher state.
+         *
+         * @param {boolean} open Pass toggle boolean back through two-way bind.
+         */
+        $scope.toggled = function (open) {
```
Comment (jeff-phillips-18) [inline src/navigation/launcher-navigation-directive.js]: We don't typically doc the controller functions.

File: src/navigation/launcher-navigation.html [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+           role="menuitem"
+           ng-repeat="item in items()">
+        <a class="pf-launcher-link" ng-href="{{item.href}}" target="_blank">
+          <span class="fa fa-3x" ng-class="item.iconClass" ng-if="!hiddenIcons()"></span>
+          <div class="pf-launcher-entry-title">{{item.title}}</div>
+        </a>
+      </div>
+    </div>
+  </div>
+</div>
```
Comment (jeff-phillips-18) [inline src/navigation/launcher-navigation.html]: Please add ending newline characters

File: src/navigation/launcher-navigation-directive.js [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+ </example>
+ */
+angular.module('patternfly.navigation').directive('pfLauncherNavigation', [
+  function () {
+
+    'use strict';
+
+    return {
+      restrict: 'A',
+      replace: true,
```
Comment (jeff-phillips-18) [inline src/navigation/launcher-navigation-directive.js]: Replace is deprecated. Should be removed.

Comment (cdcabrera) [inline src/navigation/launcher-navigation-directive.js]: can do

File: src/navigation/launcher-navigation-directive.js (hunk 3 above)
Comment (cdcabrera) [inline src/navigation/launcher-navigation-directive.js]: one-way binds. we could change them, but it's not necessary to send the info back out. unless we're planning on using it for something?

File: src/navigation/launcher-navigation-directive.js (hunk 4 above)
Comment (cdcabrera) [inline src/navigation/launcher-navigation-directive.js]: gotcha, i can remove

File: src/navigation/launcher-navigation.html (hunk 5 above)
Comment (cdcabrera) [inline src/navigation/launcher-navigation.html]: haha, can do, just removed it

**Comment by cdcabrera** [comment]
Can do on the naming as well "application-launcher"

File: src/navigation/launcher-navigation-directive.js (hunk 3 above)
Comment (jeff-phillips-18) [inline src/navigation/launcher-navigation-directive.js]: I've only ever used '&' to pass functions.

Comment (cdcabrera) [inline src/navigation/launcher-navigation-directive.js]: gotcha, if there's a need for consistency we can alter them easy enough. the limiting factor in using them would be related to directive/component loading

**Comment by dtaylor113** [comment]
> where the component/directive should live, currently placed under navigation

ok
> preference in component/directive naming

prefer 'pfApplicationLauncher'
> any recommendations on isolated scope properties

Seem ok
> have a rough of the Less/CSS was aiming to place within Patternfly, but dependent on need can place where ever recommended

Best to have Less/CSS in base Patternfly.  Is there a base PatternFly app launcher?   You can add the Less/CSS to angular-patternfly initially, then put up a PR to add the Less/CSS to base Patternfly.  Then once that is merged, another PR to remove the Less/CSS from angular-patternfly, fun huh? ;-)

**Comment by dtaylor113** [comment]
Pls be sure to squash all those commits before merging!  -thanks

**Comment by dtaylor113** [comment]
Hi, would it be possible to show the app. launcher icon on the navbar-pf black background, the same way pfNotificationDrawer does?  That is most likely where it will be placed most often.
Ex:
image

**Comment by cdcabrera** [comment]
Indeed, applying it for a demo/example on the navbar-pf background, good idea can do

**Comment by cdcabrera** [comment]
Integrating styling into base PatternFly. Once example for the app launcher is updated and base PatternFly styling integrated will convert PR to non-review

**Comment by dtaylor113** [comment]
Hi, just a heads up that OpenShift Console/Catalog is waiting on this PR in order to add the App. Launcher to the console application.  I understand that this PR is waiting on PR in base PF.

**Comment by bleathem** [comment]
PatternFly 3.24.0 has been released.  Is this PR ready to merge?

**Comment by bleathem** [comment]
@cdcabrera commented via slack that this PR is ready to merge.

**Pull request merged**